go/rpcclientlib/viewing_key_client.go

```diff
@@ -216,14 +221,40 @@ func (c *ViewingKeyClient) addFromAddressToCallParamsIfMissing(method string, ar
 		}
 	}
 
+	// We only modify `eth_call` requests where the `from` field is not set.
 	if callParams[reqJSONKeyFrom] != nil {
-		// We only modify `eth_call` requests where the `from` field is not set.
 		return args, nil
 	}
-	callParams[reqJSONKeyFrom] = c.viewingKeyAddr
 
-	args[0] = callParams
+	// We ensure that the `data` field is present.
+	if callParams[reqJSONKeyData] == nil {
+		return nil, fmt.Errorf("eth_call request did not have its `from` or its `data` field set. Aborting " +
+			"request as it will not be possible to encrypt the response")
+	}
+	data, ok := callParams[reqJSONKeyData].(string)
+	if !ok {
+		return nil, fmt.Errorf("eth_call request's `data` field was not of the expected type `string`")
+	}
 
+	// We iterate over the `data` field, looking for an argument that matches the viewing key address. If we find one,
+	// we set the `from` field to that address.
+	data = data[10:] // We remove the leading "0x" (1 bytes/2 chars) and the method ID (4 bytes/8 chars).
+	for i := 0; i < len(data); i += ethCallPaddedArgLen {
+		if i+ethCallPaddedArgLen > len(data) {
+			return nil, fmt.Errorf("eth_call request did not have its `from` field set, and its `data` field " +
+				"did not contain an address matching a viewing key. Aborting request as it will not be possible to " +
+				"encrypt the response")
+		}
+		// Each encoded parameter in the `data` field is 32 padded bytes. Since an address is 20 bytes, we only want
```

Comment:
are you sure this is the general rule?
What if you're sending: "aaaaaaaaa..."- 100 times "a" basically?
Would that be 32 bytes? Is it split up

Reply:
I believe so, but EthereumDocs is only a second language for me:

```For all types of length at most 32 bytes, the EVENT_INDEXED_ARGS array contains the value directly, padded or sign-extended (for signed integers) to 32 bytes, just as for regular ABI encoding. However, for all “complex” types or types of dynamic length, including all arrays, string, bytes and structs, EVENT_INDEXED_ARGS will contain the Keccak hash of a special in-place encoded value (see [Encoding of Indexed Event Parameters](https://docs.soliditylang.org/en/develop/abi-spec.html#indexed-event-encoding)), rather than the encoded value directly. This allows applications to efficiently query for values of dynamic-length types (by setting the hash of the encoded value as the topic), but leaves applications unable to decode indexed values they have not queried for. For dynamic-length types, application developers face a trade-off between fast search for predetermined values (if the argument is indexed) and legibility of arbitrary values (which requires that the arguments not be indexed). Developers may overcome this tradeoff and achieve both efficient search and arbitrary legibility by defining events with two arguments — one indexed, one not — intended to hold the same value.``` - https://docs.soliditylang.org/en/develop/abi-spec.html

Reply:
this is really hard to parse.

Reply:
This method might be good to have tests for perhaps? Somewhere we can add edge-cases as we discover them haha

Reply:
It's tested by the wallet extension tests. Do you mean failure tests?